File: codes/Haskell/p012.hs
{- 
 - Solution to Project Euler problem 12
 - Copyright (c) Sankha Saha. All rights reserved.
 - 
 -  
 -   
 -}

import qualified EulerLib


main = putStrLn (show ans)
ans = head $ filter (\n -> (divisors n) > 500) $ map triangleNumber [0..]

triangleNumber i = div (i * (i + 1)) 2

divisors :: Int -> Int
divisors n = (sum [1 | k <- [1 .. (EulerLib.sqrt n)], mod n k == 0]) * 2 - EulerLib.boolToInt ((EulerLib.sqrt n)^2 == n)

File: codes/mathematica/p012.m
(* 
 * Solution to Project Euler problem 12
 * Copyright (c) Sankha Saha. All rights reserved.
 * 
 *  
 *   
 *)


(* We do a straightforward search with some help from built-in functions. *)
TriangleNumber[n_] = Sum[i, {i, n}];
i = 1;
While[DivisorSigma[0, TriangleNumber[i]] <= 500, i++]
TriangleNumber[i]
